cmdstanpy/cmdstan_args.py

```diff
@@ -0,0 +1,398 @@
+"""
+CmdStan arguments
+"""
+import os
+import tempfile
+from typing import List, Union
+
+import numpy as np
+
+from cmdstanpy.utils import check_csv, read_metric
+
+
+class SamplerArgs(object):
+    """Arguments for the NUTS adaptive sampler."""
+
+    def __init__(
+        self,
+        warmup_iters: int = None,
+        sampling_iters: int = None,
+        save_warmup: bool = False,
+        thin: int = None,
+        max_treedepth: int = None,
+        metric: Union[str, List[str]] = None,
+        step_size: Union[float, List[float]] = None,
+        adapt_engaged: bool = None,
+        adapt_delta: float = None,
+    ) -> None:
+        """Initialize object."""
+        self.warmup_iters = warmup_iters
+        self.sampling_iters = sampling_iters
+        self.save_warmup = save_warmup
+        self.thin = thin
+        self.max_treedepth = max_treedepth
+        self.metric = metric
+        self.metric_file = None
+        self.step_size = step_size
+        self.adapt_engaged = adapt_engaged
+        self.adapt_delta = adapt_delta
+
+    def validate(self, chains: int) -> None:
+        """
+        Check arguments correctness and consistency.
+
+        * adaptation and warmup args are consistent
+        * if file(s) for metric are supplied, check contents.
+        * length of per-chain lists equals specified # of chains
+        """
+        if type(chains) is not int or chains < 1:
+            raise ValueError(
+                "sampler expects number of chains to be greater than 0"
+            )
+
+        if self.warmup_iters is not None:
+            if self.warmup_iters < 0:
+                raise ValueError(
+                    'warmup_iters must be a non-negative integer'.format(
+                        self.warmup_iters
+                    )
+                )
+            if self.adapt_engaged and self.warmup_iters == 0:
+                raise ValueError(
+                    'adaptation requested but 0 warmup iterations specified, '
+                    'must run warmup iterations'
+                )
+
+        if self.sampling_iters is not None:
+            if self.sampling_iters < 0:
```

Comment:
can this be 0, will it work?

Reply:
yes, it will work.

(you could do this if you wanted to see stepsize, mass matrix after a certain amount of warmup, I suppose.   the imaginative user will try anything and everything.  just matching CmdStan's constraints here.)

Reply:
Ok, good.

PyStan fails with 0 iter (only warmup) and it's on my list to-be-fixed